### PR TITLE
Add support for Azure AI Search (Azure Cognitive Search)

### LINKED ISSUE #14906: Add Azure cognitive search implementation
Just like we have an implementation for Lucene and Elasticsearch, it would be awesome to add support for [Azure cognitive search](https://azure.microsoft.com/en-us/products/ai-services/ai-search/?ef_id=_k_EAIaIQobChMIvZ7s_Z2SgwMVHgytBh3gug2zEAAYAiAAEgI6j_D_BwE_k_&OCID=AIDcmm5edswduu_SEM__k_EAIaIQobChMIvZ7s_Z2SgwMVHgytBh3gug2zEAAYAiAAEgI6j_D_BwE_k_&gad_source=1&gclid=EAIaIQobChMIvZ7s_Z2SgwMVHgytBh3gug2zEAAYAiAAEgI6j_D_BwE)

### PR DESCRIPTION
**Pull request by MikeAlhayek**
Fix #14906

## Demo of Managing Azure AI Indices
In this demo, we enabled the `Azure AI Search` feature. Then created a new index for the `Article` content type. When creating, resetting or rebuilding and index, we trigger a job in the background that will upload all existing content items to the Azure AI Search index so the data will become searchable shortly after the action.
azure-ai-search-index

## Using the Search module to search Azure AI Index
In this demo, we enabled the Search module to enable frontend search. Then used the previously created `articles` index as the default search index.
azure-ai-search-rearching

## Search Settings Improvements
All the front end search settings are now on a single page separated by tabs.

azure-cognitive-search-2

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/OrchardCore.Search.Azure.CognitiveSearch.csproj
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/OrchardCore.Search.Azure.CognitiveSearch.csproj:10]: Search Search

File: src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/Views/AzureCognitiveSearchSettings.Edit.cshtml
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/Views/AzureCognitiveSearchSettings.Edit.cshtml:9]:
```suggestion
    <div class="alert alert-warning">@T["No indice exists! Please create at least one index to configure Azure Cognitive Search service."]</div>
```

File: src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/Views/AzureCognitiveSearchSettings.Edit.cshtml
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/Views/AzureCognitiveSearchSettings.Edit.cshtml:27]:
```suggestion
    <span class="hint">@T["A comma-separated list of fields to use during search. Leave blank to search all fields."]</span>
```

File: src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/Views/_ViewImports.cshtml
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/Views/_ViewImports.cshtml:5]:
```suggestion
```

File: src/OrchardCore/OrchardCore.Search.Azure.CognitiveSearch.Core/Permissions.cs
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Search.Azure.CognitiveSearch.Core/Permissions.cs:36]: Make static

File: src/OrchardCore/OrchardCore.Search.Azure.CognitiveSearch.Core/Services/AzureCognitiveSearchDocumentManager.cs
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Search.Azure.CognitiveSearch.Core/Services/AzureCognitiveSearchDocumentManager.cs:17]: FrozenDictionary

File: src/OrchardCore/OrchardCore.Search.Azure.CognitiveSearch.Core/Services/AzureCognitiveSearchIndexManager.cs
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Search.Azure.CognitiveSearch.Core/Services/AzureCognitiveSearchIndexManager.cs:209]: Cache per tenant. IMemoryCache.

File: src/OrchardCore/OrchardCore.Search.Azure.CognitiveSearch.Core/Services/SearchClientFactory.cs
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Search.Azure.CognitiveSearch.Core/Services/SearchClientFactory.cs:27]: Lock or ConcurrentDictionary

**Review by sebastienros — APPROVED** [review]

**Comment by sebastienros** [comment]
@bleroy FYI

File: src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/AdminMenu.cs
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/AdminMenu.cs:6]: Why it's not `OrchardCore.Search.Azure` or `OrchardCore.Search.Azure.CognitiveSearch`?

File: src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/Controllers/AdminController.cs
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/Controllers/AdminController.cs:76]:
```suggestion
```

**Comment by bleroy** [comment]
Oooooh, very nice! I'll show that to the team.

File: src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/Controllers/AdminController.cs
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/Controllers/AdminController.cs:134]:
```suggestion
    public ActionResult IndexFilterPOST(AdminIndexViewModel model) =>
        RedirectToAction("Index", new RouteValueDictionary {
            { "Options.Search", model.Options.Search }
        });
```

**Comment by bleroy** [comment]
Also, there's been a rebranding: https://azure.microsoft.com/en-us/products/ai-services/ai-search/

File: src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/Controllers/AdminController.cs
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/Controllers/AdminController.cs:159]:
```suggestion

                    await _notifier.SuccessAsync(H["Indices successfully removed."]);
```

File: src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/Controllers/AdminController.cs
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/Controllers/AdminController.cs:356]: Why not the supported cultures?!!

File: src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/OrchardCore.Search.Azure.CognitiveSearch.csproj
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/OrchardCore.Search.Azure.CognitiveSearch.csproj:10]:
```suggestion
      Creates Azure Cognitive Search indexes to support search scenarios, and introduces a preconfigured container-enabled content type.
```

File: src/OrchardCore/OrchardCore.Indexing.Abstractions/DocumentIndex.cs
Comment (hishamco) [inline src/OrchardCore/OrchardCore.Indexing.Abstractions/DocumentIndex.cs:56]: Export into a new file

File: src/OrchardCore/OrchardCore.Indexing.Abstractions/DocumentIndex.cs
Comment (hishamco) [inline src/OrchardCore/OrchardCore.Indexing.Abstractions/DocumentIndex.cs:66]: Same here

File: src/docs/reference/modules/AzureCognitiveService/README.md
Comment (hishamco) [inline src/docs/reference/modules/AzureCognitiveService/README.md:25]:
```suggestion
When the `Search` module is enabled along with `Azure Cognitive Search`, you can run the frontend site search against your Azure Cognitive Search indices.
```

**Review by hishamco — CHANGES_REQUESTED** [review]

**Comment by hishamco** [comment]
> Also, there's been a rebranding: https://azure.microsoft.com/en-us/products/ai-services/ai-search/

@bleroy I just saw that while reviewing the PR, thanks for mentioning this

@MikeAlhayek seems there's another round for renaming :)

File: src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/AdminMenu.cs
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/AdminMenu.cs:6]: Azure could have many search services

**Comment by MikeAlhayek** [comment]
@bleroy 
> Also, there's been a rebranding: https://azure.microsoft.com/en-us/products/ai-services/ai-search/

Uggg. I'll rename things once I am done.

File: src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/AdminMenu.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/AdminMenu.cs:6]: Sure, while the context is search there's no confusion, moreover, after the rebranding we might use `OrchardCore.Search.AzureAI`

File: src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/Controllers/AdminController.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/Controllers/AdminController.cs:356]: FYI #14934

File: src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/AdminMenu.cs
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/AdminMenu.cs]: Not a fan of `OrchardCore.Search.AzureAI`, but can't think of a better name. So I used it.

**Comment by MikeAlhayek** [comment]
@bleroy I updated the screen casts to show the final available functionality.

However, there is something weird in the `CreateOrUpdateIndexAsync` method. I tried to update the index using the following line
```
await _searchIndexClient.CreateOrUpdateIndexAsync(indexOptions, allowIndexDowntime: true, onlyIfUnchanged: true);
```
I get an exception when a field has changed. Here is an example of an exception when I tried to change the `Content__ContentItem__DisplayText__Analyzed` field

```
Existing field 'Content__ContentItem__DisplayText__Analyzed' cannot be changed.
Status: 400 (Bad Request)
ErrorCode: OperationNotAllowed

Content:
{"error":{"code":"OperationNotAllowed","message":"Existing field 'Content__ContentItem__DisplayText__Analyzed' cannot be changed.","details":[{"code":"CannotChangeExistingField","message":"Existing field 'Content__ContentItem__DisplayText__Analyzed' cannot be changed."}]}}
```

I am not sure if this is by design or an issue? As a workaround, I delete the index and create new one every time instead of using `CreateOrUpdateIndexAsync` method as you can [see here](https://github.com/OrchardCMS/OrchardCore/blob/366e61e76917a22a64d6fac526be3e03d8e1e6c8/src/OrchardCore/OrchardCore.Search.AzureAI.Core/Services/AzureAISearchIndexManager.cs#L195-L206).

### Feedback

Creating a instance of `SearchClient` requires Index name in the second argument. It would be nicer API if the index name is provider in the methods arguments instead of the constructor.

So instead of this
```
var client = new SearchClient("endpoint", "indexName", credentials);
var searchResult = await client.SearchAsync<SearchDocument>("searchTest", searchOptions)
```

do 

```
var client = new SearchClient("endpoint", credentials);
var searchResult = await client.SearchAsync<SearchDocument>("indexName", "searchTest", searchOptions)
```

This way we do not have to create a client for each index and manage the lifetime of each clients. If you look at my implementation, I created a factory to provide me a client for each index as you can see here [SearchClientFactory](https://github.com/OrchardCMS/OrchardCore/blob/366e61e76917a22a64d6fac526be3e03d8e1e6c8/src/OrchardCore/OrchardCore.Search.AzureAI.Core/Services/SearchClientFactory.cs#L22-L34)

@sebastienros 
I'll merge this PR once we ship OC 1.8

File: src/docs/resources/owners/README.md
Comment (hishamco) [inline src/docs/resources/owners/README.md:24]: Good one :)

**Comment by bleroy** [comment]
@MikeAlhayek > I am not sure if this is by design or an issue?

This is normal. Here's an article that describes what conditions require an index rebuild: https://learn.microsoft.com/en-us/azure/search/search-howto-reindex

File: src/docs/resources/owners/README.md
Comment (MikeAlhayek) [reply src/docs/resources/owners/README.md:24]: ?

**Comment by hishamco** [comment]
@MikeAlhayek is this ready? I need to check the requested changes, then approve

**Comment by MikeAlhayek** [comment]
Yes it is ready

File: src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/Controllers/AdminController.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/Controllers/AdminController.cs]: Extra line

File: src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/Controllers/AdminController.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/Controllers/AdminController.cs]: Add a line before

**Review by hishamco — CHANGES_REQUESTED** [review]
Please don't resolve the comment unless they are really resolved, I saw few formating issues that I pointed to them at the beginning

**Comment by hishamco** [comment]
Again I dislike the current naming, my suggestions are

1. `OrcharcCore.Search.AzureAI`
2. `OrchardCore.Search.Azure`

but I'd leave this to @sebasteinros, @Piedone and other folks to decide

File: src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/Controllers/AdminController.cs
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/Controllers/AdminController.cs]: Why do you need and extra line before? Typically we add a new line before return not after every line.

**Comment by MikeAlhayek** [comment]
I am already using `OrcharcCore.Search.AzureAI` not planning to rename everything all over again

**Comment by hishamco** [comment]
@bleroy your thought about the module name?

File: src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/Controllers/AdminController.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/Controllers/AdminController.cs]: Also, after the end of any block. Please check the ASP.NET Core code conventions which we already used since the beginning

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Search.Azure.CognitiveSearch/Controllers/AdminController.cs]: @hishamco I think you are not looking at the latest code. there was a new line there already. Maybe go back and refresh the page.

**Comment by Piedone** [comment]
`OrcharcCore.Search.AzureAI` is good. The other search features use the same pattern. (And it's about "AI" search specifically.)

**Comment by hishamco** [comment]
Seems you merged without a final approval, but it's fine while the name changed :)

Thanks

**Pull request merged**
